operator/cmd/k8s_identity.go

```diff
@@ -10,78 +10,208 @@ import (
-	err := identityRateLimiter.Wait(ctx)
-	if err != nil {
-		return err
+	cell.Provide(newIdentitiesGCConfig),
+	cell.Provide(newIdentitiesGC),
+
+	// Invoke forces the instantiation of the identity gc
+	cell.Invoke(func(*IdentitiesGC) {}),
+)
+
+// IdentitiesGCConfig contains the configuration for the identity-gc.
+type IdentitiesGCConfig struct {
+	AllocationMode string
+
+	IdentityGCInterval       time.Duration
+	IdentityHeartbeatTimeout time.Duration
+	EndpointGCInterval       time.Duration
+
+	RateInterval time.Duration
+	RateLimit    int64
+
+	MetricsEnabled bool
+}
+
+func newIdentitiesGCConfig(daemonCfg *option.DaemonConfig, operatorCfg *operatorOption.OperatorConfig) IdentitiesGCConfig {
+	return IdentitiesGCConfig{
+		AllocationMode:           daemonCfg.IdentityAllocationMode,
+		IdentityGCInterval:       operatorCfg.IdentityGCInterval,
+		IdentityHeartbeatTimeout: operatorCfg.IdentityHeartbeatTimeout,
+		EndpointGCInterval:       operatorCfg.EndpointGCInterval,
+		RateInterval:             operatorCfg.IdentityGCRateInterval,
+		RateLimit:                operatorCfg.IdentityGCRateLimit,
+		MetricsEnabled:           operatorCfg.EnableMetrics,
 	}
-	err = clientset.CiliumV2().CiliumIdentities().Delete(
-		ctx,
-		identity.Name,
-		metav1.DeleteOptions{
-			Preconditions: &metav1.Preconditions{
-				UID:             &identity.UID,
-				ResourceVersion: &identity.ResourceVersion,
-			},
-		})
-	if err != nil {
-		log.WithError(err).Error("Unable to delete identity")
-	} else {
-		log.WithField(logfields.Identity, identity.GetName()).Info("Garbage collected identity")
+}
+
+// IdentityGCDeps contains all the dependencies for the identity-gc.
+// They will be provided through dependency injection.
+type IdentityGCDeps struct {
+	cell.In
+
+	Logger logrus.FieldLogger
+	LC     hive.Lifecycle
+
+	Clientset        k8sClient.Clientset
+	IdentityResource resource.Resource[*v2.CiliumIdentity]
+}
+
+func newIdentitiesGC(deps IdentityGCDeps, cfg IdentitiesGCConfig) *IdentitiesGC {
+	obj := &IdentitiesGC{
+		cfg:              cfg,
+		logger:           deps.Logger,
+		clientset:        deps.Clientset,
+		identityResource: deps.IdentityResource,
+		heartbeatStore: identity.NewIdentityHeartbeatStore(
+			cfg.IdentityHeartbeatTimeout,
+		),
+		rateLimiter: rate.NewLimiter(
+			cfg.RateInterval,
+			cfg.RateLimit,
+		),
 	}
 
-	return err
+	deps.LC.Append(obj)
+
+	return obj
 }
 
-func updateIdentity(ctx context.Context, clientset k8sClient.Clientset, identity *v2.CiliumIdentity) error {
-	_, err := clientset.CiliumV2().CiliumIdentities().Update(
-		ctx,
-		identity,
-		metav1.UpdateOptions{})
-	if err != nil {
-		log.WithError(err).Error("Updating Identity")
-	} else {
-		log.WithField(logfields.Identity, identity.GetName()).Debug("Updated identity")
+// IdentitiesGC represents the Cilium identities periodic GC.
+type IdentitiesGC struct {
+	logger logrus.FieldLogger
+
+	cfg IdentitiesGCConfig
+
+	clientset        k8sClient.Clientset
```

Comment:
using the subclient here might be cleaner? e.g.  interface returned by `clientset.CiliumV2().CliiumIdentities()`.